src/library/Cardano/CoinSelection/Fee.hs

```diff
@@ -305,13 +300,14 @@ reduceChangeOutputs threshold (Fee totalFee) changeOutputs
             [] -> []
   where
     payFee :: (Fee, Coin) -> Coin
-    payFee (Fee f, Coin c) = Coin (c - f)
```

Comment:
I am not sure how we ended up with this function in the first place. This has probably slipped through review and refactoring. On the original code, this used to be:

```hs
-- | Reduce single change output by a given fee amount. If fees are too big for
-- a single coin, returns a `Coin 0`.
reduceSingleChange :: (Fee, Coin) -> Coin
reduceSingleChange (Fee fee, Coin chng)
    | chng >= fee =
          Coin (chng - fee)
    | otherwise =
          Coin 0
```

And the case of underflow was actually properly handled. It now is anyway, but it still feels a lot heavier than the guard above and the use a mere `Natural`.

Reply:
@KtorZ wrote:
> I am not sure how we ended up with this function in the first place. This has probably slipped through review and refactoring.

Quite likely.

Though to my mind, this really does support the case for having a type that _doesn't_ support unchecked subtraction.

If the underlying type implements `Num` (as `Natural` does), it becomes easy for a future programmer to _unintentionally_ introduce a regression just by using `-` in an unchecked context.

At least with `SafeNatural`, the attack surface is greatly reduced, as now we only have:

```hs
sub      :: SafeNatural -> SafeNatural -> Maybe SafeNatural
distance :: SafeNatural -> SafeNatural ->       SafeNatural
```

In my opinion, being able to rely on the type system in this way is _much_ safer than relying on code review to spot the use of an unchecked subtraction. (Of course, this doesn't negate the need for good code review.)

> it still feels a lot heavier than the guard above and the use a mere Natural.

After change 79fdc177d677ad1974ef426b836a0e644fd4a25e, this is now similar in spirit to the original:
```hs
    payFee :: (Fee, Coin) -> Coin                                                                                                                                                     
    payFee (Fee f, Coin c) = Coin $                                                                                                                                                   
        fromMaybe SN.zero (c `SN.sub` f)   
```
Which should be similar (in spirit) to the original.